File: examples/python/simple_example.py
"""
Simple example demonstrating task-queue usage.

Run this example after starting a worker:
    python -m taskqueue worker --queues=default
"""

from taskqueue import TaskQueue
import time

# Initialize queue
queue = TaskQueue(broker='redis://localhost:6379/0')

# Define a simple task
@queue.task
def add(x, y):
    """Add two numbers."""
    return x + y

@queue.task
def process_data(data):
    """Process data with a delay."""
    print(f"Processing data: {data}")
    time.sleep(1)
    return {"processed": len(data), "sum": sum(data)}

# Define a task with retry
@queue.task(retries=3, retry_backoff=True)
def flaky_task(attempt):
    """Task that might fail."""
    print(f"Attempt {attempt}")
    if attempt < 3:
        raise Exception("Not ready yet!")
    return "Success!"

# Define a high-priority task
@queue.task(queue='high_priority', priority=0)
def urgent_task(message):
    """Urgent task that should be processed first."""
    print(f"URGENT: {message}")
    return message.upper()

if __name__ == "__main__":
    print("Enqueuing tasks...")

    # Simple task
    task1 = add.delay(5, 3)
    print(f"Task 1 ID: {task1.id}")

    # Task with complex arguments
    data = list(range(100))
    task2 = process_data.delay(data)
    print(f"Task 2 ID: {task2.id}")

    # Urgent task
    task3 = urgent_task.delay("Process this immediately!")
    print(f"Task 3 ID: {task3.id}")

    # Flaky task (will retry)
    task4 = flaky_task.delay(1)
    print(f"Task 4 ID: {task4.id}")

    print("\nWaiting for results...")

    # Get results
    result1 = task1.get(timeout=10)
    print(f"Result 1: {result1}")

    result2 = task2.get(timeout=10)
    print(f"Result 2: {result2}")

    result3 = task3.get(timeout=10)
    print(f"Result 3: {result3}")

    result4 = task4.get(timeout=30)
    print(f"Result 4: {result4}")

    print("\nAll tasks completed!")

File: examples/go/simple_example.go
// Simple example demonstrating task-queue usage in Go
//
// Run this example after starting a worker:
// go run main.go

package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/task-queue/go/taskqueue"
)

// Task handlers
func add(ctx context.Context, x, y int) (interface{}, error) {
	log.Printf("Adding %d + %d", x, y)
	return x + y, nil
}

func processData(ctx context.Context, data []int) (interface{}, error) {
	log.Printf("Processing %d items", len(data))
	sum := 0
	for _, v := range data {
		sum += v
	}
	time.Sleep(1 * time.Second)
	return map[string]interface{}{
		"processed": len(data),
		"sum":       sum,
	}, nil
}

func urgentTask(ctx context.Context, message string) (interface{}, error) {
	log.Printf("URGENT: %s", message)
	return message, nil
}

func flakyTask(ctx context.Context, attempt int) (interface{}, error) {
	log.Printf("Attempt %d", attempt)
	if attempt < 3 {
		return nil, fmt.Errorf("not ready yet!")
	}
	return "Success!", nil
}

func main() {
	// Create queue
	queue := taskqueue.NewQueue(
		"default",
		"redis://localhost:6379/0",
	)

	fmt.Println("Enqueuing tasks...")

	// Simple task
	task1, err := queue.Enqueue("add", 5, 3)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Task 1 ID: %s\n", task1.ID)

	// Task with complex arguments
	data := make([]int, 100)
	for i := range data {
		data[i] = i
	}
	task2, err := queue.Enqueue("processData", data)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Task 2 ID: %s\n", task2.ID)

	// Urgent task
	task3, err := queue.Enqueue(
		"urgentTask",
		"Process this immediately!",
		taskqueue.WithQueue("high_priority"),
		taskqueue.WithPriority(0),
	)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Task 3 ID: %s\n", task3.ID)

	// Flaky task (will retry)
	task4, err := queue.Enqueue(
		"flakyTask",
		1,
		taskqueue.WithRetries(3),
		taskqueue.WithRetryBackoff(true),
	)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Task 4 ID: %s\n", task4.ID)

	fmt.Println("\nWaiting for results...")

	// Get results with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	result1, err := task1.Get(ctx)
	if err != nil {
		log.Printf("Task 1 failed: %v", err)
	} else {
		fmt.Printf("Result 1: %v\n", result1)
	}

	result2, err := task2.Get(ctx)
	if err != nil {
		log.Printf("Task 2 failed: %v", err)
	} else {
		fmt.Printf("Result 2: %v\n", result2)
	}

	result3, err := task3.Get(ctx)
	if err != nil {
		log.Printf("Task 3 failed: %v", err)
	} else {
		fmt.Printf("Result 3: %v\n", result3)
	}

	ctx4, cancel4 := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel4()
	result4, err := task4.Get(ctx4)
	if err != nil {
		log.Printf("Task 4 failed: %v", err)
	} else {
		fmt.Printf("Result 4: %v\n", result4)
	}

	fmt.Println("\nAll tasks completed!")
}
